istioctl/cmd/vm-bootstrap.go

```diff
@@ -0,0 +1,630 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"crypto"
+	"crypto/sha256"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/user"
+	"path"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/spf13/cobra"
+	"golang.org/x/crypto/ssh"
+	"golang.org/x/crypto/ssh/knownhosts"
+	"golang.org/x/crypto/ssh/terminal"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+
+	networking "istio.io/api/networking/v1alpha3"
+
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pkg/config/schema/collections"
+	"istio.io/istio/pkg/spiffe"
+	"istio.io/istio/security/pkg/k8s/secret"
+	"istio.io/istio/security/pkg/pki/util"
+)
+
+var (
+	all               bool
+	certDuration      time.Duration
+	dumpDir           string
+	istioProxyImage   string
+	mutualTLS         bool
+	organization      string
+	remoteDirectory   string
+	scpPath           string
+	spiffeTrustDomain string
+	sshAuthMethod     ssh.AuthMethod
+	sshKeyLocation    string
+	sshIgnoreHostKeys bool
+	sshUser           string
+	startIstio        bool
+	workloadNamespace string
+
+	workloadKind = collections.IstioNetworkingV1Alpha3Workloadentries.Resource().GroupVersionKind()
+)
+
+type workloadEntryAddressKeys struct {
+	CaCert           []byte
+	Cert             []byte
+	Key              []byte
+	ServiceName      string
+	ServiceNamespace string
+}
+
+type describedCert struct {
+	PemEncodedCa []byte
+	Ca           *x509.Certificate
+	Key          crypto.PrivateKey
+}
+
+type remoteResponse struct {
+	typ     uint8
+	message string
+}
+
+func fetchSingleWorkloadEntry(workloadName string, client model.ConfigStore) ([]model.Config, string, error) {
+	workloadSplit := strings.Split(workloadName, ".")
+	if len(workloadSplit) != 2 {
+		return nil, "", fmt.Errorf("workload name: %s is not in the format: workloadName.Namespace", workloadName)
+	}
+
+	we := client.Get(workloadKind, workloadSplit[0], workloadSplit[1])
+	if we == nil {
+		return nil, "", fmt.Errorf("workload entry: %s in namespace: %s was not found", workloadSplit[0], workloadSplit[1])
+	}
+
+	return []model.Config{*we}, workloadSplit[1], nil
+}
+
+func fetchAllWorkloadEntries(client model.ConfigStore) ([]model.Config, string, error) {
+	list, err := client.List(workloadKind, workloadNamespace)
+	return list, workloadNamespace, err
+}
+
+func getCertificate(kubeClient kubernetes.Interface) (describedCert, error) {
+	secret, err := kubeClient.CoreV1().Secrets(istioNamespace).Get(context.TODO(), secret.CASecret, metav1.GetOptions{})
```

Comment:
This is a very, very bad security practice - and should not be part of istioctl. 

We can add an interface to allow certain authenticated users to get signed certs for other identities if we really want to automate provisioning - but please don't ex-filtrate the private keys as a matter of routine. 

Many of the tools to manage VMs also support provisioning certificate - with proper intermediate and without exposing the root.

Reply:
And I know we use as example the security/tool that does exactly the same - but it's meant as example, while istioctl is usually assumed to be the official tool.

Reply:
I agree it's unhygenic. Happy to make it an endpoint somewhere, but let's still realize that doesn't actually fix the scenario your describing here. Even if we make it an endpoint that signs for effectively the wide range of addresses we want to sign for, even if we make it a new intermediate.

If we don't trust the operator machine that needs to be able to do this bootstraping action. Then no matter what we're going to have problems. Because what's stopping someone from reading the token used to talk to this endpoint that generates certs. There's gotta be a bottom rung for the operator, and if we can't trust that than we have bigger problems.

This was just meant as a first stab (def not the most ideal end state) that we can hack on, and happy to change to using an intermediate, but regardless of where this ends up the scenario you described could always in theory happen.

Reply:
That's what audit trails are for - as well as policies. 

If you exfiltrate the root private key - you can sign anything you want without any trace. 

If it's an API you must call - passing your own identity as admin and a CSR, with a server-side config indicating you are allowed to impersonate or create provisioning certs:

- the server can send audit messages to record what you did
- the server can restrict what certs you provision - for example only in your own namespace

In majority of cases, having a policy like ns-admin-sa.namespace cert allowed to sign/generate
*.namespace certificates is pretty safe ( with the audit included ), and doesn't expose the entire mesh if one admin machine is owned. 

And it works even with CI/CD systems - where you can restrict the CI/CD to just one namespace and provision machines for that NS only.